### PR TITLE
Handle multiple concurrent D-Bus calls of the app

### PR DESCRIPTION
**Pull request by GabrielNagy**
This commit reworks the lifecycle management of the application in a separate package, moving away from a context cancelling approach in favor of handling multiple runs using waitgroups.

The app mutex is now the responsibility of the `lifecycle` package as well, abstracting all of the functionality outside of the app package.

To manage the lifecycle of the app, we attach a `Lifecycle` instance to the `proxyManagerBus` object.

We call `Start()` when a client calls our D-Bus function, which increments the lifecycle counter and causes subsequent calls to block until the current one finishes.

When a run finishes we call `RunDone()` which signals the lifecycle to decrement the counter and start processing the next run if applicable.

Once the counter reaches zero, the `Wait()` function returns a joined representation of all errors that occurred during the runs, thus effectively ending the program.

DEENG-602

### CODE REVIEW

File: internal/app/export_test.go [hunk 1]
```diff
@@ -14,11 +14,13 @@ type MockAuthorizer struct {
 
 // MockProxy is a mock proxy.
 type MockProxy struct {
-	ApplyError bool
+	ApplyCount   int
+	ApplyError   bool
+	SleepOnApply time.Duration
 }
 
 // IsSenderAllowed is a mock implementation of authorizerer, returning an error if requested in the mock.
```
Comment (denisonbarbosa) [inline internal/app/export_test.go:22]:
```suggestion
// IsSenderAllowed is a mock implementation of Authorizer, returning an error if requested in the mock.
```

File: internal/app/export_test.go [hunk 2]
```diff
@@ -14,11 +14,13 @@ type MockAuthorizer struct {
...
-func (m *MockAuthorizer) IsSenderAllowed(ctx context.Context, action string, sender dbus.Sender) (err error) {
+func (m *MockAuthorizer) IsSenderAllowed(action string, sender dbus.Sender) (err error) {
```
Comment (denisonbarbosa) [inline internal/app/export_test.go:23]: Usually, when I see this kind of function name, I expect a answer to a question (i.e. `true | false`) instead of an error. Maybe we can thing of a better name for it? Something in the likes of `CheckSender`? (I suck at naming, but I'm sure we can find a good name with three brains working together)

File: internal/authorizer/authorizer.go [hunk 3]
```diff
@@ -90,7 +89,7 @@ func New(bus *dbus.Conn, args ...option) *Authorizer {
 // IsSenderAllowed returns nil if the user is allowed to perform a given operation.
 // Based on the D-Bus sender it will query the user's credentials and then
 // attempt to authorize the action using polkit.
-func (a Authorizer) IsSenderAllowed(ctx context.Context, action string, sender dbus.Sender) (err error) {
+func (a Authorizer) IsSenderAllowed(action string, sender dbus.Sender) (err error) {
```
Comment (denisonbarbosa) [inline internal/authorizer/authorizer.go:92]: Same thing on the naming here

File: internal/authorizer/authorizer.go [hunk 4]
```diff
@@ -109,11 +108,11 @@ func (a Authorizer) IsSenderAllowed(ctx context.Context, action string, sender d
 		return errors.New("can't get pid from dbus credentials")
 	}
 
-	return a.isAllowed(ctx, action, pid, uid)
+	return a.isAllowed(action, pid, uid)
 }
 
 // isAllowed returns nil if the given uid/pid are allowed to perform the given action.
-func (a Authorizer) isAllowed(ctx context.Context, action string, pid uint32, uid uint32) (err error) {
+func (a Authorizer) isAllowed(action string, pid uint32, uid uint32) (err error) {
```
Comment (denisonbarbosa) [inline internal/authorizer/authorizer.go:115]: Ditto on the naming

**Review by denisonbarbosa — APPROVED** [review]
I like it a lot! Dropping the need of specifying contexts whilst also being concurrency safe is so nice. One minor suggestion on a function naming (also, sorry for not spotting it before :smiling_face_with_tear: ), but everything else LGTM.

File: internal/lifecycle/lifecycle.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+	return &Lifecycle{
+		timeout: timeout,
+		quitted: make(chan struct{}),
+		started: make(chan struct{}),
+	}
+}
+
+// Start starts or extends the lifecycle one run and blocks the next run until
+// the current run finishes.
+func (l *Lifecycle) Start() {
```
Comment (didrocks) [inline internal/lifecycle/lifecycle.go:44]: I feel Start() to be a little bit ambiguous. as it’s not the Start of the Lifecycle object, but that we have a new run going. I think we should have something like AddRun() and DoneRun() (you have RunDone() which was already similar)

File: internal/lifecycle/lifecycle.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+// - quit to be explicitly requested
+// returning a joined representation of all errors that occurred during the runs.
+func (l *Lifecycle) Wait() error {
+	timer := time.NewTimer(l.timeout)
+
+	for {
+		select {
+		case <-l.started:
+			l.counter.Wait()
+			return errors.Join(l.errors...)
```
Comment (didrocks) [inline internal/lifecycle/lifecycle.go:73]: 1.2000000000000000000000000000 :)

File: internal/lifecycle/lifecycle.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+// QuitRequested returns true if the application has been requested to quit.
+func (l *Lifecycle) QuitRequested() bool {
+	l.quitMu.RLock()
+	defer l.quitMu.RUnlock()
+
+	return l.exitRequested
+}
+
+func (l *Lifecycle) increment() {
```
Comment (didrocks) [inline internal/lifecycle/lifecycle.go:100]: You are calling increment/decrement only once, I would just inline in the parent function, wdyt?

File: internal/lifecycle/lifecycle.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+	exitRequested bool
+	quitted       chan struct{}
+	started       chan struct{}
+}
+
+// New returns a new Lifecycle with the given timeout.
+func New(timeout time.Duration) *Lifecycle {
+	return &Lifecycle{
+		timeout: timeout,
+		quitted: make(chan struct{}),
```
Comment (didrocks) [inline internal/lifecycle/lifecycle.go:37]: No really thought about it fully yet: but would it be possible to merged started/quitted by using the `nil` value smartly? (Remember that we can check for a nil channel, listening on it will be undefinitively)

File: internal/app/export_test.go (hunk 2 above)
Comment (didrocks) [reply internal/app/export_test.go:23]: This is true.

`CheckSenderAllowed` maybe (as we don’t check for the sender validity, but the fact that the sender can perform that action). We need to update adsys too then!

File: internal/app/app.go [hunk 9]
```diff
@@ -53,32 +49,34 @@ type options struct {
...
-	b.mu.Lock()
-	defer b.mu.Unlock()
-
 	log.Debugf("Sender %s called Apply(%q, %q, %q, %q, %q, %q)", sender, http, https, ftp, socks, no, mode)
 
-	// We need to cancel the context in a deferred function to get the final
-	// state of the error variable, and to let the main thread know it's safe to quit.
+	// Application was already asked to quit, so return an error without applying anything
+	if b.l.QuitRequested() {
+		return dbus.MakeFailedError(errors.New("application exit requested, cannot apply proxy settings"))
```
Comment (didrocks) [inline internal/app/app.go:64]: I’m a little bit not at ease with this: it means that we could get there, BUT, a new request has just come. So, we can proceed, no?

**Review by didrocks — CHANGES_REQUESTED** [review]
First, congrats on the lifecycle implementation, I think you had fun implementing it :)

I know you implemented it like a graceful shutdown pattern, but I think here, we can simplify it a little bit for our use case. I find it a little bit sad that we get to some current "new Apply call" (see my first comment), but we refuse doing it because a shutdown was already requested, while we got there, so we can theorically proceed it.

Also, that makes me realize another thing: we can only accept on Apply at a time, otherwise, there is a race in the generation of the files on disk and we can end up with mix configuration of 2 Apply calls.

So basically, our Apply is a queue of requests (and we should queue any new "Apply" as they arrive and treat them one after another). Finally, we can tear the whole process down only when the queue is empty. I think, coming from this prerequisite will simplify the process a lot, even if what you wrote is correct from different hypothesis and well executed :)

(Also, if in this new model you keep a lifecycle package, dedicated tests in this package would be appreciated :)).

**Comment by didrocks** [comment]
> But doesn't this achieve the exact same functionality I already implemented?

Not exactly, here the requests are running in parallel now, aren’t they? (maybe I missed something).

The fact that we take a "lock" only when entering the request will vastly simplify the code, but I’m not opposed to merge it if you confirm the above is working as expected. It’s a corner case as you say, but corner case will happen :p So, this is why I’m in favor of simplifying the code as much as we can when possible. But again, the functionality is small enough that we can get to this in a second stage. (just that the tests written for lifecycle package itself, which, even if covered by the rest, still need some dedicated one to declare the contract of this package, will then be useless).

So again, it’s up to you on this, you only have my advice here :)
